File: lib/features/rooms/services/room_service.dart
import 'package:garbanzo_ai/core/api_client.dart';
import 'package:garbanzo_ai/features/chat/models/thinking_level.dart';
import 'package:garbanzo_ai/features/rooms/models/room_models.dart';

/// REST client for multi-person chat rooms.
class RoomService {
  RoomService._();
  static final RoomService instance = RoomService._();

  final _api = ApiClient.instance;

  Future<List<Room>> listRooms({int page = 1, int pageSize = 50}) async {
    final resp = await _api.get(
      '/api/v1/rooms',
      queryParameters: {'page': page, 'page_size': pageSize},
    );
    if (resp.statusCode != 200) {
      throw Exception('Failed to list rooms: ${resp.statusCode}');
    }
    final items = (resp.data['items'] as List)
        .map((e) => Room.fromJson(e as Map<String, dynamic>))
        .toList();
    return items;
  }

  Future<Room> createRoom({
    required String name,
    String? description,
    bool isPublic = false,
    int maxAgentTurnDepth = 3,
    List<String> memberEmails = const [],
  }) async {
    final resp = await _api.post(
      '/api/v1/rooms',
      data: {
        'name': name,
        'description': ?description,
        'is_public': isPublic,
        'max_agent_turn_depth': maxAgentTurnDepth,
        'member_emails': memberEmails,
      },
    );
    if (resp.statusCode != 201 && resp.statusCode != 200) {
      throw Exception('Failed to create room: ${resp.statusCode} ${resp.data}');
    }
    return Room.fromJson(resp.data as Map<String, dynamic>);
  }

  Future<Room> getRoom(String roomId) async {
    final resp = await _api.get('/api/v1/rooms/$roomId');
    if (resp.statusCode != 200) {
      throw Exception('Failed to load room: ${resp.statusCode}');
    }
    return Room.fromJson(resp.data as Map<String, dynamic>);
  }

  Future<Room> updateRoom(
    String roomId, {
    String? name,
    String? description,
    bool? isPublic,
    int? maxAgentTurnDepth,
    String? ownerId,
  }) async {
    final payload = <String, dynamic>{};
    if (name != null) payload['name'] = name;
    if (description != null) payload['description'] = description;
    if (isPublic != null) payload['is_public'] = isPublic;
    if (maxAgentTurnDepth != null) {
      payload['max_agent_turn_depth'] = maxAgentTurnDepth;
    }
    if (ownerId != null) payload['owner_id'] = ownerId;
    final resp = await _api.patch('/api/v1/rooms/$roomId', data: payload);
    if (resp.statusCode != 200) {
      throw Exception('Failed to update room: ${resp.statusCode}');
    }
    return Room.fromJson(resp.data as Map<String, dynamic>);
  }

  Future<void> deleteRoom(String roomId) async {
    final resp = await _api.delete('/api/v1/rooms/$roomId');
    if (resp.statusCode != 204) {
      throw Exception('Failed to delete room: ${resp.statusCode}');
    }
  }

  Future<RoomMember> addMember(
    String roomId,
    String email, {
    String role = 'member',
  }) async {
    final resp = await _api.post(
      '/api/v1/rooms/$roomId/members',
      data: {'user_id': email, 'role': role},
    );
    if (resp.statusCode != 201) {
      throw Exception('Failed to add member: ${resp.statusCode} ${resp.data}');
    }
    return RoomMember.fromJson(resp.data as Map<String, dynamic>);
  }

  Future<void> removeMember(String roomId, String email) async {
    final resp = await _api.delete('/api/v1/rooms/$roomId/members/$email');
    if (resp.statusCode != 204) {
      throw Exception('Failed to remove member: ${resp.statusCode}');
    }
  }

  /// Mute or unmute the current user's notifications for [roomId].
  ///
  /// [duration] is one of `8h`, `1w`, `forever`, `unmute` — validated
  /// server-side by the `RoomMuteUpdate` schema.
  Future<RoomMember> setMute(String roomId, String duration) async {
    final resp = await _api.patch(
      '/api/v1/rooms/$roomId/members/me/mute',
      data: {'duration': duration},
    );
    if (resp.statusCode != 200) {
      throw Exception('Failed to update mute: ${resp.statusCode} ${resp.data}');
    }
    return RoomMember.fromJson(resp.data as Map<String, dynamic>);
  }

  Future<RoomAgent> addAgent(
    String roomId, {
    required String name,
    required String model,
    String provider = 'ollama',
    String? avatar,
    String? systemPrompt,
    ThinkingLevel? thinkingLevel,
    String responseMode = 'mention',
    int turnOrder = 0,
    bool isModerator = false,
    List<String>? enabledTools,
  }) async {
    final resp = await _api.post(
      '/api/v1/rooms/$roomId/agents',
      data: {
        'name': name,
        'model': model,
        'provider': provider,
        'avatar': ?avatar,
        'system_prompt': ?systemPrompt,
        'thinking_level': ?thinkingLevel?.name,
        'response_mode': responseMode,
        'turn_order': turnOrder,
        'is_active': true,
        'is_moderator': isModerator,
        'enabled_tools': ?enabledTools,
      },
    );
    if (resp.statusCode != 201) {
      throw Exception('Failed to add agent: ${resp.statusCode} ${resp.data}');
    }
    return RoomAgent.fromJson(resp.data as Map<String, dynamic>);
  }

  Future<RoomAgent> updateAgent(
    String roomId,
    String agentId,
    Map<String, dynamic> patch,
  ) async {
    final resp = await _api.patch(
      '/api/v1/rooms/$roomId/agents/$agentId',
      data: patch,
    );
    if (resp.statusCode != 200) {
      throw Exception('Failed to update agent: ${resp.statusCode}');
    }
    return RoomAgent.fromJson(resp.data as Map<String, dynamic>);
  }

  Future<void> deleteAgent(String roomId, String agentId) async {
    final resp = await _api.delete('/api/v1/rooms/$roomId/agents/$agentId');
    if (resp.statusCode != 204) {
      throw Exception('Failed to delete agent: ${resp.statusCode}');
    }
  }

  Future<List<RoomMessage>> listMessages(
    String roomId, {
    int page = 1,
    int pageSize = 100,
  }) async {
    final resp = await _api.get(
      '/api/v1/rooms/$roomId/messages',
      queryParameters: {'page': page, 'page_size': pageSize},
    );
    if (resp.statusCode != 200) {
      throw Exception('Failed to list messages: ${resp.statusCode}');
    }
    return (resp.data['items'] as List)
        .map((e) => RoomMessage.fromJson(e as Map<String, dynamic>))
        .toList();
  }

  Future<String> exportMarkdown(String roomId) async {
    final resp = await _api.get(
      '/api/v1/rooms/$roomId/export',
      queryParameters: {'format': 'markdown'},
    );
    if (resp.statusCode != 200) {
      throw Exception('Failed to export: ${resp.statusCode}');
    }
    return resp.data.toString();
  }

  Future<List<Room>> search(
    String query, {
    String scope = 'all',
    int page = 1,
    int pageSize = 20,
  }) async {
    final resp = await _api.get(
      '/api/v1/rooms/search',
      queryParameters: {
        'q': query,
        'scope': scope,
        'page': page,
        'page_size': pageSize,
      },
    );
    if (resp.statusCode != 200) {
      throw Exception('Search failed: ${resp.statusCode}');
    }
    return (resp.data['items'] as List)
        .map((e) => Room.fromJson(e as Map<String, dynamic>))
        .toList();
  }
}

File: backend/app/services/room_service.py
"""Service for Room CRUD + member + agent management."""

from __future__ import annotations

import logging
import uuid
from dataclasses import dataclass

from sqlalchemy import desc, func, or_, select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from app.models.friendship import Friendship
from app.models.room import Room, RoomAgent, RoomMember, RoomMessage
from app.models.user import User
from app.services.mute_util import resolve_mute_until

logger = logging.getLogger(__name__)


class RoomPermissionError(Exception):
    """Raised when a user is not allowed to perform an action."""


class RoomNotFoundError(Exception):
    """Raised when a room does not exist or the user can't see it."""


class UnknownUserError(Exception):
    """Raised when a referenced user email does not exist."""

    def __init__(self, emails: list[str]):
        self.emails = emails
        super().__init__(f"Unknown user(s): {', '.join(emails)}")


@dataclass
class RoomSearchHit:
    room: Room
    is_member: bool


class RoomService:
    def __init__(self, db: AsyncSession):
        self.db = db

    # ------------------------------------------------------------------ Rooms

    async def create(
        self,
        owner_id: str,
        name: str,
        description: str | None = None,
        is_public: bool = False,
        max_agent_turn_depth: int = 3,
        mode: str = "chat",
        member_emails: list[str] | None = None,
    ) -> Room:
        room_id = str(uuid.uuid4())
        room = Room(
            id=room_id,
            name=name,
            description=description,
            owner_id=owner_id,
            is_public=is_public,
            max_agent_turn_depth=max_agent_turn_depth,
            mode=mode,
        )
        # Dedup invitees, excluding owner
        invitees: list[str] = []
        seen = {owner_id}
        for email in member_emails or []:
            if email in seen:
                continue
            seen.add(email)
            invitees.append(email)

        # Validate invitee emails exist in users table before inserting
        if invitees:
            existing = set(
                (await self.db.execute(select(User.email).where(User.email.in_(invitees))))
                .scalars()
                .all()
            )
            missing = [e for e in invitees if e not in existing]
            if missing:
                raise UnknownUserError(missing)

        # Friend-graph privacy guard: a block between the owner and an
        # invitee makes that invitee unaddable, with a message that doesn't
        # disclose the block (or who placed it).
        for email in invitees:
            if await self._blocked_pair_exists(owner_id, email):
                raise RoomPermissionError(f"Unable to add {email} to the room.")

        self.db.add(room)
        self.db.add(RoomMember(room_id=room_id, user_id=owner_id, role="owner"))
        for email in invitees:
            self.db.add(RoomMember(room_id=room_id, user_id=email, role="member"))

        await self.db.commit()
        return await self._get_with_relations(room_id)

    async def _get_with_relations(self, room_id: str) -> Room:
        result = await self.db.execute(
            select(Room)
            .where(Room.id == room_id)
            .options(
                selectinload(Room.members).selectinload(RoomMember.user),
                selectinload(Room.agents),
            )
        )
        return result.scalar_one()

    async def get(self, room_id: str, *, viewer_id: str | None = None) -> Room | None:
        query = (
            Room.active()
            .where(Room.id == room_id)
            .options(
                selectinload(Room.members).selectinload(RoomMember.user),
                selectinload(Room.agents),
            )
        )
        room = (await self.db.execute(query)).scalar_one_or_none()
        if room is None:
            return None
        # Refresh cached relationships so any in-session mutations (new
        # members, agent edits) are visible.
        await self.db.refresh(room, attribute_names=["members", "agents"])
        if (
            viewer_id is not None
            and not room.is_public
            and not await self.is_member(room_id, viewer_id)
        ):
            return None
        return room

    async def is_member(self, room_id: str, user_id: str) -> bool:
        row = (
            await self.db.execute(
                select(RoomMember.user_id).where(
                    RoomMember.room_id == room_id, RoomMember.user_id == user_id
                )
            )
        ).first()
        return row is not None

    async def list_for_user(
        self, user_id: str, page: int = 1, page_size: int = 20
    ) -> tuple[list[Room], int]:
        base = (
            Room.active()
            .join(RoomMember, RoomMember.room_id == Room.id)
            .where(RoomMember.user_id == user_id)
        )
        count_query = select(func.count()).select_from(base.subquery())
        total = (await self.db.execute(count_query)).scalar() or 0

        query = (
            base.options(selectinload(Room.members), selectinload(Room.agents))
            .order_by(desc(Room.updated_at))
            .offset((page - 1) * page_size)
            .limit(page_size)
        )
        rooms = list((await self.db.execute(query)).scalars().unique().all())
        return rooms, total

    async def search(
        self,
        viewer_id: str,
        query: str,
        scope: str = "all",
        page: int = 1,
        page_size: int = 20,
    ) -> tuple[list[RoomSearchHit], int]:
        """Search rooms by name or description.

        Scope:
          - ``mine`` — only rooms the user is a member of
          - ``public`` — only public rooms (any owner)
          - ``all`` (default) — member rooms + public rooms
        """
        normalized = (query or "").strip()
        if not normalized:
            return [], 0
        escaped = normalized.replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_")
        pattern = f"%{escaped}%"

        member_rooms_subq = select(RoomMember.room_id).where(RoomMember.user_id == viewer_id)

        base = Room.active().where(
            or_(
                Room.name.ilike(pattern, escape="\\"),
                Room.description.ilike(pattern, escape="\\"),
            )
        )
        if scope == "mine":
            base = base.where(Room.id.in_(member_rooms_subq))
        elif scope == "public":
            base = base.where(Room.is_public == True)  # noqa: E712
        else:  # "all"
            base = base.where(or_(Room.id.in_(member_rooms_subq), Room.is_public == True))  # noqa: E712

        count_query = select(func.count()).select_from(base.subquery())
        total = (await self.db.execute(count_query)).scalar() or 0
        if total == 0:
            return [], 0

        query_stmt = (
            base.options(selectinload(Room.members), selectinload(Room.agents))
            .order_by(desc(Room.updated_at))
            .offset((page - 1) * page_size)
            .limit(page_size)
        )
        rooms = list((await self.db.execute(query_stmt)).scalars().unique().all())

        member_ids = set((await self.db.execute(member_rooms_subq)).scalars().all())
        hits = [RoomSearchHit(room=r, is_member=r.id in member_ids) for r in rooms]
        return hits, total

    async def update(
        self,
        room_id: str,
        user_id: str,
        *,
        name: str | None = None,
        description: str | None = None,
        is_public: bool | None = None,
        max_agent_turn_depth: int | None = None,
        mode: str | None = None,
        owner_id: str | None = None,
    ) -> Room:
        room = await self._require_owner(room_id, user_id)

        if name is not None:
            room.name = name
        if description is not None:
            room.description = description
        if is_public is not None:
            room.is_public = is_public
        if max_agent_turn_depth is not None:
            room.max_agent_turn_depth = max_agent_turn_depth
        if mode is not None:
            room.mode = mode
        if owner_id is not None and owner_id != room.owner_id:
            # New owner must be a member already
            new_owner_row = (
                await self.db.execute(
                    select(RoomMember).where(
                        RoomMember.room_id == room_id,
                        RoomMember.user_id == owner_id,
                    )
                )
            ).scalar_one_or_none()
            if new_owner_row is None:
                raise RoomPermissionError("New owner must already be a member of the room")
            # Demote old owner, promote new owner
            old_owner_row = (
                await self.db.execute(
                    select(RoomMember).where(
                        RoomMember.room_id == room_id,
                        RoomMember.user_id == room.owner_id,
                    )
                )
            ).scalar_one_or_none()
            if old_owner_row is not None:
                old_owner_row.role = "member"
            new_owner_row.role = "owner"
            room.owner_id = owner_id

        await self.db.commit()
        return await self._get_with_relations(room_id)

    async def delete(self, room_id: str, user_id: str) -> None:
        room = await self._require_owner(room_id, user_id)
        room.is_deleted = True
        await self.db.commit()

    # ----------------------------------------------------------------- Members

    async def _blocked_pair_exists(self, a: str, b: str) -> bool:
        """Whether a friend-graph block exists between two users (either
        direction). Rooms respect blocks without caring who placed them."""
        row = (
            await self.db.execute(
                select(Friendship.id).where(
                    Friendship.status == "blocked",
                    or_(
                        (Friendship.requester_email == a) & (Friendship.addressee_email == b),
                        (Friendship.requester_email == b) & (Friendship.addressee_email == a),
                    ),
                )
            )
        ).scalar_one_or_none()
        return row is not None

    async def add_member(
        self, room_id: str, user_id: str, new_user_id: str, role: str = "member"
    ) -> RoomMember:
        await self._require_owner(room_id, user_id)
        if await self._blocked_pair_exists(user_id, new_user_id):
            # Same non-disclosing message as at room creation.
            raise RoomPermissionError(f"Unable to add {new_user_id} to the room.")
        existing = (
            await self.db.execute(
                select(RoomMember).where(
                    RoomMember.room_id == room_id, RoomMember.user_id == new_user_id
                )
            )
        ).scalar_one_or_none()
        if existing:
            return existing
        member = RoomMember(room_id=room_id, user_id=new_user_id, role=role)
        self.db.add(member)
        await self.db.commit()
        return member

    async def remove_member(self, room_id: str, user_id: str, target_user_id: str) -> None:
        room = await self._require_owner(room_id, user_id)
        if target_user_id == room.owner_id:
            raise RoomPermissionError("Cannot remove the owner; transfer ownership first")
        row = (
            await self.db.execute(
                select(RoomMember).where(
                    RoomMember.room_id == room_id,
                    RoomMember.user_id == target_user_id,
                )
            )
        ).scalar_one_or_none()
        if row is None:
            return
        await self.db.delete(row)
        await self.db.commit()

    async def list_members(self, room_id: str) -> list[RoomMember]:
        rows = (
            (
                await self.db.execute(
                    select(RoomMember)
                    .where(RoomMember.room_id == room_id)
                    .options(selectinload(RoomMember.user))
                    .order_by(RoomMember.joined_at)
                )
            )
            .scalars()
            .all()
        )
        return list(rows)

    async def set_mute(self, room_id: str, user_id: str, duration: str) -> RoomMember:
        """Mute or unmute room notifications for ``user_id``.

        ``duration`` is one of ``"8h"``, ``"1w"``, ``"forever"``, ``"unmute"``
        (validated by the ``MuteUpdate`` schema at the API boundary).
        """
        member = (
            await self.db.execute(
                select(RoomMember).where(
                    RoomMember.room_id == room_id, RoomMember.user_id == user_id
                )
            )
        ).scalar_one_or_none()
        if member is None:
            raise RoomNotFoundError("Not a room member")

        member.muted_until = resolve_mute_until(duration)

        await self.db.commit()
        await self.db.refresh(member)
        return member

    # ------------------------------------------------------------------ Agents

    async def add_agent(
        self,
        room_id: str,
        user_id: str,
        *,
        name: str,
        model: str,
        provider: str = "ollama",
        avatar: str | None = None,
        system_prompt: str | None = None,
        thinking_level: str | None = None,
        response_mode: str = "mention",
        turn_order: int = 0,
        is_active: bool = True,
        is_moderator: bool = False,
        enabled_tools: list[str] | None = None,
    ) -> RoomAgent:
        await self._require_owner(room_id, user_id)
        agent = RoomAgent(
            id=str(uuid.uuid4()),
            room_id=room_id,
            name=name,
            avatar=avatar,
            provider=provider,
            model=model,
            system_prompt=system_prompt,
            thinking_level=thinking_level,
            response_mode=response_mode,
            turn_order=turn_order,
            is_active=is_active,
            is_moderator=is_moderator,
            enabled_tools=enabled_tools,
        )
        self.db.add(agent)
        await self.db.commit()
        await self.db.refresh(agent)
        return agent

    async def update_agent(
        self,
        room_id: str,
        user_id: str,
        agent_id: str,
        **fields,
    ) -> RoomAgent:
        await self._require_owner(room_id, user_id)
        agent = (
            await self.db.execute(
                select(RoomAgent).where(RoomAgent.id == agent_id, RoomAgent.room_id == room_id)
            )
        ).scalar_one_or_none()
        if agent is None:
            raise RoomNotFoundError("Agent not found")
        # None means "clear" for nullable columns (the endpoint only forwards
        # fields the client actually sent); other fields ignore None.
        nullable = {"avatar", "system_prompt", "thinking_level", "enabled_tools"}
        for key, value in fields.items():
            if value is None and key not in nullable:
                continue
            setattr(agent, key, value)
        await self.db.commit()
        await self.db.refresh(agent)
        return agent

    async def delete_agent(self, room_id: str, user_id: str, agent_id: str) -> None:
        await self._require_owner(room_id, user_id)
        agent = (
            await self.db.execute(
                select(RoomAgent).where(RoomAgent.id == agent_id, RoomAgent.room_id == room_id)
            )
        ).scalar_one_or_none()
        if agent is None:
            return
        await self.db.delete(agent)
        await self.db.commit()

    async def list_agents(self, room_id: str) -> list[RoomAgent]:
        rows = (
            (
                await self.db.execute(
                    select(RoomAgent)
                    .where(RoomAgent.room_id == room_id)
                    .order_by(RoomAgent.turn_order, RoomAgent.created_at)
                )
            )
            .scalars()
            .all()
        )
        return list(rows)

    # ---------------------------------------------------------------- Messages

    async def list_messages(
        self, room_id: str, page: int = 1, page_size: int = 50
    ) -> tuple[list[RoomMessage], int]:
        base = select(RoomMessage).where(RoomMessage.room_id == room_id)
        count_query = select(func.count()).select_from(base.subquery())
        total = (await self.db.execute(count_query)).scalar() or 0

        rows = (
            (
                await self.db.execute(
                    base.order_by(RoomMessage.created_at)
                    .offset((page - 1) * page_size)
                    .limit(page_size)
                )
            )
            .scalars()
            .all()
        )
        return list(rows), total

    async def all_messages_for_export(self, room_id: str) -> list[RoomMessage]:
        rows = (
            (
                await self.db.execute(
                    select(RoomMessage)
                    .where(RoomMessage.room_id == room_id)
                    .order_by(RoomMessage.created_at)
                )
            )
            .scalars()
            .all()
        )
        return list(rows)

    # --------------------------------------------------------------- Internals

    async def _require_owner(self, room_id: str, user_id: str) -> Room:
        room = (
            await self.db.execute(
                Room.active()
                .where(Room.id == room_id)
                .options(selectinload(Room.members), selectinload(Room.agents))
            )
        ).scalar_one_or_none()
        if room is None:
            raise RoomNotFoundError("Room not found")
        if room.owner_id != user_id:
            raise RoomPermissionError("Only the room owner can perform this action")
        return room
